### ISSUE TITLE
Expand function thoughts

### ISSUE THREAD

**Issue by justin1121** [OpenMined/OpenMined#205]
I have spent some time working through the expand function and have encountered a few issues that conflict with how we are currently doing things. In the pytorch documentation it says that to expand a tensor without allocating any new memory all you need to do is set the stride of the expanded dimensions to zero. While I'm not 100% sure how this works mathematically it feels intuitively correct because to get to the next element in a dimension you just wouldn't stride anywhere. I've written the code that does the above but then I ran into some conflicts with how we're currently doing things.

1. Size member of Tensor class

Currently the Size member of the tensor class is calculated by multiplying all the shapes together. So should we be updating the Size when a tensor gets expanded? The main issue with this is that its not clear if the Size should equal multiplying the shapes together or is it the number of elements in the underlying array. There are a few places its used to iterate through the underlying array. This would have to change if we change the Size when we expand the tensor because then the Size would cause you to go out of bounds if the Size is being used to iterate through the array.

2. IsContiguous

Currently the IsContiguous function just checks to see if the last dimension's stride is equal to 1. If you expanded a {4, 1} tensor into a {4, 4} the strides would look like {1, 0} and this would cause IsContiguous to fail. Maybe that is expected behaviour? Or we could modify IsContiguous to handle this case?

3. Dividing by strides

There are couple places where we are dividing by strides. If a stride was zero this could cause a crash. This is done in the reduce functionality and the GetIndex function. I'm not sure what the solution is. Are there other equations that do the same thing without dividing by the stride?

I could be totally off base with how I'm approaching this function. Let me know your thoughts! 

**Issue**
https://github.com/OpenMined/PySyft/issues/475

**PyTorch Docs**
http://pytorch.org/docs/master/tensors.html#torch.Tensor.expand

**Comment by emregoynugur** [comment]
We thought of utilizing `shape`, `strides`, and `isContiguous` like PyTorch does, when we first implemented the transpose function. But for some reason we preferred keeping tensors contiguous. Since we adopted a fast development process, we never had a chance to discuss this issue again :). 

I honestly haven't had the time to check the `View` function, but that might also be wrong (at least in the inline case). It looks like it only changes the `shape` and does not update `strides` at all. So, thanks for bringing this up!

1.  If we still want to keep the current implementation, I think `size` should return the total number of elements (in this case the expanded number of elements). However, I am aware that `expand` does not allocate new memory, so we will need to do some modifications in any case.

2. I think it is the expected behaviour, since the expanded result is just a view and the actual data is not copied. However, as far as I know tensors in our implementations are always contiguous. If I am wrong and they can be non-contiguous, all of the element-wise op implementations are wrong, because they don't check if the tensors are contiguous.

3. If we are going to adopt this approach, this would only be one of the necessary modifications :).

The only thing I am sure is that `IsContiguous` is working as intended :) We either start making use of it or come up with another solution.

**Comment by justin1121** [comment]
> If we still want to keep the current implementation, I think size should return the total number of elements (in this case the expanded number of elements).

So for example, a {4, 1} matrix extended to a {4, 4} matrix, its size would become 16. I think this makes sense we would just have to be careful with using the size to iterate through the tensor's data.

> I think it is the expected behaviour, since the expanded result is just a view and the actual data is not copied.

So expanding a tensor makes it no longer contiguous. Are there some other functions that might have the same effect?

**Comment by iamtrask** [comment]
So expanding a tensor makes it no longer contiguous. Are there some other functions that might have the same effect?

Yes - several. Concatenation is the first one that comes to mind.

**Comment by iamtrask** [comment]
index-select as well i believe

**Comment by emregoynugur** [comment]
We can focus on the implementation of the operators in this list (https://github.com/onnx/onnx/blob/master/docs/Operators.md) for the moment. I think `View` and few others are not there at the moment, however this would allow us to import models from other frameworks, which already implement all the operators in Python. We can start by importing simple models.

If we want to support View in our framework, then lets discuss how we should modify/extend existing ops. I think `transpose` would especially benefit from this change.

**Comment by justin1121** [comment]
I like this approach. That list seems like a pretty good place to start. I think a bit more direction in which functions to implement first is needed. I'm not really sure what functions are most important. I would be willing to follow that list. Any ideas how we could host/track that list ourselves?

In regards to `View`, I'm not 100% what the current issue is. I know `View` isn't currently checking if the tensor is contiguous which is a problem if we support non-contiguous tensors. Perhaps thats the main issue here. How we want to support non-contiguous tensors. I noticed `Squeeze` is in the list above. Our current implementation relies on `View`. I'm sure `Squeeze` could be implemented in a different way but it would probably be similar to what `View` already does. Interestingly, I just did some experimentation and it looks like `Squeeze` supports non-contiguous tensors while `View` does not. So, I think we'd have to reimplement `Squeeze` in a different way.

**Comment by iamtrask** [comment]
My two cents - I think we definitely want to be able to support View, Squeeze, and onnx conversion. It sounds like it's possible to do all 3. Fixing the strides bug in View is a must (i believe that was my mistake). This would allow us to do Expand without copying data, yes? Seems like the best route to go.

**Comment by iamtrask** [comment]
As far as a roadmap - sounds like step one would be to upgrade all our other operations with an "isContiguous" check - don't have to implement the non-contiguous versions right away... just have an if statement that requires a "Clone()" to convert it to be contiguous before proceeding.

**Comment by justin1121** [comment]
I think this is a good starting point. But I think the Contiguous() function would be what we want not a Clone(). Need some sort of logic to turn a non-contiguous into contiguous. A Clone() would just copy the data wouldn't it? 

Would the user be responsible for calling the Clone()/Contiguous() or would the function do it for them? We could just throw an exception for now?

**Comment by justin1121** [comment]
If the above is true, I can get started working on this. Think we would need to implement the Contiguous() function as well. Let me know if you have any thoughts on how that should be done, because I'm not sure.

**Comment by emregoynugur** [comment]
> If the above is true, I can get started working on this. Think we would need to implement the Contiguous() function as well. Let me know if you have any thoughts on how that should be done, because I'm not sure.

We probably have to implement `Contiguous()`. The simplest way would be to utilize `GetIndex` function like our current `Transpose` implementation. However, I don't think this is the most efficient implementation method :). 

> I think this is a good starting point. But I think the Contiguous() function would be what we want not a Clone(). Need some sort of logic to turn a non-contiguous into contiguous. A Clone() would just copy the data wouldn't it?

> Would the user be responsible for calling the Clone()/Contiguous() or would the function do it for them? We could just throw an exception for now?

I think User shouldn't care about `Contiguous()`. I checked what PyTorch is doing and it seems like their `Copy()` function, which is called by their `NewClone() function, always creates a contiguous array (as much as I understand). 

See line 190 and 182.
https://github.com/torch/torch7/blob/2186e414ad8fc4dfc9f2ed090c0bf8a0e1946e62/lib/TH/generic/THTensor.c

They have different Copy functions, but check line 69 here.
https://github.com/torch/torch7/blob/6ec27dc9d941d90eea421e1bb1d0e953d68279b6/lib/TH/generic/THTensorCopy.c 

I think this part is a bit hard to follow, because of macros. See `TH_TENSOR_APPLY2_D`.
https://github.com/torch/torch7/blob/3864f3782a74715e35d98176f9014a74524a868a/lib/TH/THTensorApply.h

**Comment by justin1121** [comment]
Interesting! That TH_TENSOR_APPLY2_D is very hard to follow! I think the THTensor.c is pretty handy though to figure out how some of the contiguous related functions work.
